op-chain-ops/crossdomain/legacy_withdrawal.go

```diff
@@ -0,0 +1,116 @@
+package crossdomain
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum-optimism/optimism/op-bindings/predeploys"
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+)
+
+// LegacyWithdrawal represents a pre bedrock upgrade withdrawal.
+type LegacyWithdrawal struct {
+	Target *common.Address
+	Sender *common.Address
+	Data   []byte
+	Nonce  *big.Int
+}
+
+var _ WithdrawalMessage = (*LegacyWithdrawal)(nil)
+
+// NewLegacyWithdrawal will construct a LegacyWithdrawal
+func NewLegacyWithdrawal(target, sender *common.Address, data []byte, nonce *big.Int) *LegacyWithdrawal {
+	return &LegacyWithdrawal{
+		Target: target,
+		Sender: sender,
+		Data:   data,
+		Nonce:  nonce,
+	}
+}
+
+// Encode will serialze the Withdrawal in the legacy format so that it
+// is suitable for hashing. This assumes that the message is being withdrawn
+// through the standard optimism cross domain messaging system by hashing in
+// the L2CrossDomainMessenger address.
+func (w *LegacyWithdrawal) Encode() ([]byte, error) {
+	enc, err := EncodeCrossDomainMessageV0(w.Target, w.Sender, w.Data, w.Nonce)
+	if err != nil {
+		return nil, err
+	}
+
+	out := make([]byte, len(enc)+len(predeploys.L2CrossDomainMessengerAddr.Bytes()))
+	copy(out, enc)
+	copy(out[len(enc):], predeploys.L2CrossDomainMessengerAddr.Bytes())
+	return out, nil
+}
+
+// Decode will decode a serialized LegacyWithdrawal
+func (w *LegacyWithdrawal) Decode(data []byte) error {
```

Comment:
I'm not sure how critical this is given that this isn't exposed to much hostile input, but this will panic if `data` is not long enough.

Reply:
Ah good catch, might as well add checks for that